---
title: "Class 18: Investigating cancer genomics datasets"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Connect to the NCI-GDC database

The NCI-GDC (cancer genomics commons) is a major portal for cancer genomics data.

Here we use the bioconductor package **GenomicDataCommons** to talk to this database and download data etc. all from R. The first thing we are going to have to do is install new packages in the console using: `BiocManager::install( c("GenomicDataCommons", "TCGAbiolinks", "maftools") )`

Once installed, we must  input it into the library to use it:
```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Now lets check on GDC status to see if we can talk to the databse:
```{r}
status()
```

## Querying the GDC from R

The are four main sets of metadata that we can query, namely `projects()`, `cases()`, `files()`, and `annotations()`. We will start with `projects()`:
```{r}
projects <- getGDCprojects()
head(projects)
```

Then we will pull all the information about the cases and group the cases by the project type:
```{r}
cases_by_project <- cases() %>%
  # %>% is the pipe operator in R that would take a vector and do the function you write after it which in this case its taking the cases function on it then running the facet and then aggregations on that data. This is clearer to read because you can read it like a sentence
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
# doc-count is the number of cases there are and the FM-AD is the projects
```

## Plot this overview data

Make a barplot of the information we just made:
```{r}
x <- cases_by_project$project.project_id
head(x)
barplot(x$doc_count)
```

Now try and plot it with log:
```{r}
barplot(x$doc_count, names.arg = x$key, log="y", las=2, col = "lightblue")
```

```{r}
# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec[x$key == "TCGA-PAAD"] <- "red"

# Plot with 'log' for y axis and rotate labels with 'las'
barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)
```

## Variant Analysis with R

Find files avaliable for PAAD project that we can then examine variants from:
```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
head(maf.file)
```

Read in the MAF file we found:
```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```

Now make a summary plot of the data:
```{r}
plotmafSummary(vars)
```

## Oncoplots

A very useful summary representation of this data can be obtained via so-called oncoplots, also known as waterfall plots.

Make an oncoplot of the data:
```{r}
oncoplot(maf = vars, top = 10)
```

We can also send our plot to a PNG or PDF plot device directly, for example:
```{r}
# Oncoplot for our top 10 most frequently mutated genes
pdf("oncoplot_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
dev.off()
```

## Part 2:

```{r}
library(bio3d)
```

```{r}
seqs <- read.fasta("lecture18_sequences.fa")
#aln <- seqaln(seqs)
#aln
# we dont have muscle installed to run this code
```

Next we calculate identity per equivalent (i.e. aligned) position and then use this information to find non identical sites that do not contain gaps (i.e. indels):
```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

We are using the gap positions here and this is not what we want. So we need to exclude the gap positions:
```{r}
gaps <-gap.inspect(seqs)
gaps$f.inds
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]
mutant.sites
```
Make names for the outputs:
```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

Find the -8 before and the +8 after (i.e. all possible 9-mer sequences):
```{r}
start.position <- mutant.sites -8
end.position <- mutant.sites +8
start.position
end.position
```

Finding what the amino acids are:
```{r}
#aln$ali[2,start.position[1]:end.position[1]]
```

An example of finding the mutant sites:
```{r}
x <- c(1:6)
y <- c(1:4)
x %in% y
```

This is the code for the mutant 9-mers labeled with their name:
```{r}
# write.fasta(seqs = store.seqs, ids = mutant.names, file = "mutantseqs.fa", gap = FALSE)
```

